### ISSUE TITLE
GAS COST for `CALL` 

### ISSUE THREAD

**Issue by rkart14** [wolflo/evm-opcodes#6]
Hi Alex,  

I will try to explain my point of view on the gas costs of `CALL` and hopefully, you can somehow make me see where I could be wrong.

So, before calling a function,  EVM has to execute the following opcodes.

```
GAS
CALL
```

It's really easy to look into the source code and see what `GAS` opcode does. It's constant gas is 2 and it's operation is `opGas` quickly seen in the jump table.

`opGas` function is super simple and it does the following:

```
callContext.stack.push(new(uint256.Int).SetUint64(callContext.contract.Gas))
```

So, after the `GAS` opcode got executed, in the stack, we have the remaining gas stored. `callContext.contract.Gas` is just a variable that decreases after each operation gets executed.  So,  first point here is that it's not going to store `available_gas - 40` on the stack as you said. It's directly stores `callContext.contract.Gas`.  The reason it doesn't store `available_gas - 40` is that it didn't yet move to the `CALL` operation. So, long story, short, we have `callContext.contract.Gas` stored on the stack. Let's say this number now is 4000...

Now, `CALL` opcode comes into play. Here are what it has:

```
execute:     opCall,
constantGas: params.CallGasFrontier,
dynamicGas:  gasCall,
```

EVM first runs `constantGas` which will make `contract.Gas` => `contract.Gas - 40` because `CallGasFrontier` is 40.  So, it will become 4000-40 = 3960.

Then it runs `dynamicGas` which has the corresponding function `gasCall`.. 

`gasCall` now creates a new variable called `gas` and sums it up. (If the account is new, it adds some gas costs,  if address doesn't exist, it adds some gas costs and so on) + it also adds up memory costs.  After this, we got the final `gas` counted and it calls:

```
evm.callGasTemp, err = callGas(evm.chainRules.IsEIP150, contract.Gas, gas, stack.Back(0)) 

// callGasTemp is the most important one, because this is exactly how much gas we pass to the callee.

// The above line calls the below code, where `contract.Gas` is `availableGas`(3960), 
// `gas` is the `base` , and `stack.Back(0)` is what we put on the stack a while ago which is 4000.

availableGas = availableGas - base
gas := availableGas - availableGas/64
// If the bit length exceeds 64 bit we know that the newly calculated "gas" for EIP150
// is smaller than the requested amount. Therefore we return the new gas instead
// of returning an error.
if !callCost.IsUint64() || gas < callCost.Uint64() {
   return gas, nil
}

```

Now, if we calculate the `gas` above, it will always be less than the requested gas from the `GAS` opcode above which is 4000. Always and Always.

**Question 1.** So, I don't understand why we return the minimal from the requested gas and all_but_one_64th, because `all_but_one_64` will always be less.

**Question 2.** Unfrotunatelly, I couldn't see where it reduces contract's gas by 700. It still does by 40. Shouldn't it be 700 ?

**Comment by wolflo** [comment]
This seems to be a continuation of a conversation I am not a part of, so I may be missing some context here.

To answer your question about where `700` gets deducted, it is the updated constant gas cost for a `CALL` operation as of the Tangerine Whistle hardfork. This increase was implemented in [EIP-150](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md). You can see where the constant gas is updated in `jump_table.go` [here](https://github.com/ethereum/go-ethereum/blob/54735a67239d19d3f29e3f8ebd348a6c88ed31fa/core/vm/jump_table.go#L182).

The reason we send `all_but_one_64th` with the call is that contracts don't have to run `... GAS CALL`. That is, they can request as much or as little gas as they want. For calls to other contracts, Solidity defaults to requesting to send all remaining gas which is why this `GAS CALL` sequence is so common. But, you can request less gas than you have remaining as well, in which case we want to send the amount you requested rather than `all_but_one_64th`.

Basically, the design decision was that calls should never fail for trying to send more gas than they have. But, treating the requested call gas as a maximum makes it quite difficult to ensure that a contract sends *exactly* X amount of gas in situations where it matters ([EIP-1930](https://github.com/ethereum/EIPs/issues/1930) has some discussion of this).


It may also be relevant to mention that the gas cost for `CALL` will change again in the upcoming Berlin hardfork with [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929).

**Comment by rkart14** [comment]
Thanks for the follow-up..

Basically, from my explanation, I can presume that when contract's function calls another contract's function,  we will always send `all_but_one_64th` other than requested gas(which is all the available gas). So, 

Question 1 : I don't understand why we have 

```
if !callCost.IsUint64() || gas < callCost.Uint64() {
   return gas, nil
}
```

Because this `if` statement will always be true. 

Question 2.

`The reason we send all_but_one_64th with the call is that contracts don't have to run ... GAS CALL. That is, they can request as much or as little gas as they want` . What do you mean by `they don't have to run ... GAS CALL` ? If one contract calls another, there will be always `GAS CALL` .

**Comment by wolflo** [comment]
Well, that's not really true. We always need a requested gas value on top of the stack when `CALL` is executed, but it doesn't necessarily need to be the result of the `GAS` opcode. That's what solidity chooses to do when calling another contract, but it's not what solidity does in all cases and it's certainly not what any compiler or contract *needs* to do.
For example:
```
PUSH1 0x00
PUSH1 0x00
PUSH1 0x00
PUSH1 0x00
PUSH1 0x00
CALLER
GAS
CALL
```
```
PUSH1 0x00
PUSH1 0x00
PUSH1 0x00
PUSH1 0x00
PUSH1 0x00
CALLER
PUSH1 0x01
CALL
```

Both of these make valid calls to `msg.sender`. The first tries to send all remaining gas, resulting in `all_but_one_64th` being sent with the call. The second tries to send 1 gas. If there is 1000 gas remaining when the `CALL` operation is hit, this call will send exactly 1 gas.

In the geth `callGas()` function, `gas` is the maximum amount of gas that *can* be sent, and `callCost` is the amount that the contract *wants* to be sent. If `can >= want`, then we send the requested amount.

**Comment by rkart14** [comment]
```
evm.callGasTemp, err = callGas(evm.chainRules.IsEIP150, contract.Gas, gas, stack.Back(0))
```

`contract.Gas` we pass above is equal to `contract.Gas - 700`. Since static cost was 700 and it has already been deducted from it ,since when executing instructions, first it executes static, then dynamic, then operation itself. 

`gas` above in the parameter is the dynamic gas already counted.(memory cost + newaddress cost, address not exist cost and so on).

Below we can see that we do the following:

`availableGas = availableGas - base` 

Which means that `contract.Gas - dynamicGasCost` and this seems to be the maximum amount contract can send.  

Problem 1: What I don't understand is that everything seems to change (from 40 to 700). so why does it still do `availableGas-availableGas/64` ? Could you show an example in terms of numbers ?

Problem 2: If the `Gas` opcode is called before `CALL` , it seems like that below code's `if` statement will ALWAYS be true, because `callCost` is the gas remaining before we even deducted static and dynamic cost from the `contract.Gas`. Don'y you think ?

```
func callGas(isEip150 bool, availableGas, base uint64, callCost *uint256.Int) (uint64, error) {

  availableGas = availableGas - base
  gas := availableGas - availableGas/64
  // If the bit length exceeds 64 bit we know that the newly calculated "gas" for EIP150
  // is smaller than the requested amount. Therefore we return the new gas instead
  // of returning an error.
  if !callCost.IsUint64() || gas < callCost.Uint64() {
	  return gas, nil
  }

}
```

**Comment by wolflo** [comment]
Yes, exactly. If the `CALL` is preceded by `GAS`, the call will always send `all_but_one_64th` of the remaining gas.

But the `CALL` opcode is not always preceded by `GAS`.

**Comment by rkart14** [comment]
What I don't understand is that everything seems to change (from 40 to 700). so why does it still do `availableGas-availableGas/64` ? Could you show an example in terms of numbers ?

**Comment by wolflo** [comment]
I don't understand this question, sorry. The `min(gasRequested, availableGas - availableGas/64)` was introduced *because* of the change in the constant cost of the `CALL` opcode from 40 to 700.

**Comment by rkart14** [comment]
I think , I can ask a better question...


If the ethereum has been updated in terms of EIP150, it means that nodes ran this function:

```
func newTangerineWhistleInstructionSet() JumpTable {
	instructionSet := newHomesteadInstructionSet()
	instructionSet[BALANCE].constantGas = params.BalanceGasEIP150
	instructionSet[EXTCODESIZE].constantGas = params.ExtcodeSizeGasEIP150
	instructionSet[SLOAD].constantGas = params.SloadGasEIP150
	instructionSet[EXTCODECOPY].constantGas = params.ExtcodeCopyBaseEIP150
	instructionSet[CALL].constantGas = params.CallGasEIP150
	instructionSet[CALLCODE].constantGas = params.CallGasEIP150
	instructionSet[DELEGATECALL].constantGas = params.CallGasEIP150
	return instructionSet
}
```

So, even when someone calls old contract's code, the transaction would go to this updated EVM and would use 700 instead of 40. 

Maybe, what this is all about is that Ethereum wanted to give nodes the choice to not update their EVM's... Is that it ?

**Comment by wolflo** [comment]
There's two points I think might help here:

# 
First, the contracts that EIP-150 wanted to avoid breaking were not executing `GAS CALL`. That would have resulted in an `OUT_OF_GAS` exception on every call before EIP-150. They were executing something similar to this:
```
PUSH1 0x2b
GAS
SUB     // GAS - 43
CALL
```
Again this was not the case for every contract, but this general pattern was prevalent at the time. This means before Tangerine Whistle they were sending as much gas as possible with the call without running out of gas when the constant cost of the call was deducted. Because before Tangerine Whistle this `min(gasRequested, availableGas - availableGas/64)` did not exist. Every call sent exactly `gasRequested` or threw an `OUT_OF_GAS` exception.

So when the constant cost of a call was increased to 700, even contracts that gave themselves some buffer (say, `PUSH1 0x64 GAS SUB CALL`) would run out of gas.  

This is more of an interesting historical anecdote than anything, though. It's not necessarily relevant to understanding how the evm works now.

# 
The second point is about how hardforks work. There is no concept of an "old" contract. All code in the evm is executed according to the current set of rules. But, clients need to be able to support all previous sets of rules, because if you try to execute a transaction that was valid before the Tangerine Whistle hardfork according to the Istanbul evm semantics, it might violate some new rule and prevent you from syncing the chain.

It seems like you are trying to make sense of a pre-EIP-150 world by executing transactions with the post-EIP-150 evm semantics. This `availableGas` expression did not exist before EIP-150 (https://github.com/ethereum/go-ethereum/blob/54735a67239d19d3f29e3f8ebd348a6c88ed31fa/core/vm/gas.go#L38-L47).

To take your example, running `GAS CALL` before EIP-150 would have resulted in this `callGas()` function just returning exactly the value it was passed. Then, the evm would have tried to consume 3998 gas even though only 3958 gas remained, and it would have run out of gas here: https://github.com/ethereum/go-ethereum/blob/54735a67239d19d3f29e3f8ebd348a6c88ed31fa/core/vm/interpreter.go#L271-L274
